File: ValidationTool.Client/ValidationTool/config/absolutePaths.py
from pathlib import Path
import sys


APP_NAME = "ValidationTool"


# ============================================================
# ENVIRONMENT DETECTION
# ============================================================

def _is_frozen() -> bool:
    """
    Returns True if running from a packaged executable.
    Works for PyInstaller / bundled Python apps.
    """
    return getattr(sys, "frozen", False)


def _find_dev_root() -> Path | None:
    """
    Walk upward from this file until we find the repository root marker.
    Returns None if not found.
    """
    p = Path(__file__).resolve()

    for parent in p.parents:
        if (parent / ".validation_root").exists():
            return parent

    return None


def _get_runtime_root() -> Path:
    """
    Root folder for tools when running in deployed/runtime mode.
    Assumes the executable/tool folder is the root.
    """
    return Path(sys.executable).resolve().parent if _is_frozen() else Path.cwd()


def get_tools_root() -> Path:
    """
    Returns the root folder where tool scripts / batch files live.

    Priority:
    1. If in dev and .validation_root exists -> use repo root
    2. Otherwise -> use runtime/exe directory
    """
    dev_root = _find_dev_root()
    if dev_root is not None:
        return dev_root

    return _get_runtime_root()


# ============================================================
# DATA ROOT (ALWAYS IN DOCUMENTS)
# ============================================================

def get_documents_root() -> Path:
    """
    Base user data folder:
        Documents/ValidationTool
    """
    return Path.home() / "Documents" / APP_NAME


TOOLS_ROOT = get_tools_root()
DATA_ROOT = get_documents_root()

# Ensure data structure exists
DATA_ROOT.mkdir(parents=True, exist_ok=True)


# ============================================================
# TOOL PATHS
# ============================================================

CLIENT_PATH = TOOLS_ROOT / "ValidationTool.Client"
UI_PATH = TOOLS_ROOT / "ValidationTool.UI"
SCENESGEN_PATH = TOOLS_ROOT / "ValidationTool.ScenesGen"

CONFIG_DIR = CLIENT_PATH / "config"
HEADLESS_DIR = CLIENT_PATH / "ValidationTool" / "misc_tools" / "headless"


# ============================================================
# DATA PATHS
# ============================================================

ARTISTS_DIR = DATA_ROOT / "Artists"
REPORTS_DIR = DATA_ROOT / "Reports"
LOGS_DIR = DATA_ROOT / "Logs"

File: ValidationTool.UI/Services/Config/absolutePaths.cs
using System;
using System.IO;

namespace ValidationTool.UI.Services.Config {
    public static class Paths {//mild copy from the backend python version
        private const string APP_NAME = "ValidationTool";

        // ============================================================
        // ENVIRONMENT DETECTION
        // ============================================================

        private static string FindDevRoot() {
            var dir = new DirectoryInfo(AppContext.BaseDirectory);

            while (dir != null) {
                var marker = Path.Combine(dir.FullName, ".validation_root");
                if (File.Exists(marker) || Directory.Exists(marker))
                    return dir.FullName;

                dir = dir.Parent;
            }

            return null;
        }

        private static string GetRuntimeRoot() {
            // In deployed mode, assume tool root = folder where the exe lives
            // Could have put this inside the GetToolsRoot, but might change the
            // Hierarchy at some point, so it seemed fitting and cleaner.
            return AppContext.BaseDirectory;
        }

        private static string GetToolsRoot() {
            var devRoot = FindDevRoot();
            if (!string.IsNullOrEmpty(devRoot))
                return devRoot;

            return GetRuntimeRoot();
        }

        private static string GetDocumentsRoot() {
            return Path.Combine(
                Environment.GetFolderPath(
                    Environment.SpecialFolder.MyDocuments), APP_NAME
            );
        }

        // ============================================================
        // ROOTS
        // ============================================================

        public static readonly string TOOLS_ROOT = GetToolsRoot();
        public static readonly string DATA_ROOT = GetDocumentsRoot();

        // ============================================================
        // TOOL PATHS
        // ============================================================

        public static readonly string CLIENT_DIR =
            Path.Combine(TOOLS_ROOT, "ValidationTool.Client");

        public static readonly string UI_DIR =
            Path.Combine(TOOLS_ROOT, "ValidationTool.UI");

        public static readonly string SCENESGEN_DIR =
            Path.Combine(TOOLS_ROOT, "ValidationTool.ScenesGen");

        public static readonly string CONFIG_DIR =
            Path.Combine(CLIENT_DIR, "Config");

        public static readonly string HEADLESS =
            Path.Combine(CLIENT_DIR, "ValidationTool", "misc_tools", "headless");

        public static readonly string GEN_CONFIGS =
            Path.Combine(DATA_ROOT, "Config");

        // ============================================================
        // DATA PATHS (ALWAYS IN DOCUMENTS)
        // ============================================================

        public static readonly string ARTISTS_DIR =
            Path.Combine(DATA_ROOT, "Artists");

        public static readonly string REPORTS_DIR =
            Path.Combine(DATA_ROOT, "Reports");

        public static readonly string LOGS_DIR =
            Path.Combine(DATA_ROOT, "Logs");//empty for now, but will be using it in the future for sure

        public static readonly string RUNS_MAYA =
            Path.Combine(REPORTS_DIR, "maya_reports.json");

        public static readonly string RUNS_BLENDER =
            Path.Combine(REPORTS_DIR, "blender_reports.json");

        // ============================================================
        // INIT
        // ============================================================

        static Paths() {
            Directory.CreateDirectory(DATA_ROOT);
            Directory.CreateDirectory(ARTISTS_DIR);
            Directory.CreateDirectory(REPORTS_DIR);
            Directory.CreateDirectory(LOGS_DIR);
        }
    }
}